_source/_posts/2022-01-13-spring-cloud-contract.md

```diff
@@ -0,0 +1,893 @@
+
+Create a `Hat.java` file that is the data model for the hats the producer will serve.
+
+`src/main/java/com/example/contract/Hat.java`
+
+```java
+package com.example.contract;
+
+import lombok.AllArgsConstructor;
+import lombok.Data;
+
+@Data
+@AllArgsConstructor
+public class Hat {
+    private Long id;
+    private String name;
+    private Long size;
+    private String color;
+}
+```
+
+The hat has four attributes, an `id` as well as `name`, `size`, and `color`. Notice the two annotations. These are what Lombok helped us with. The `@Data` annotation generates getters and setters as well as a `toString()` and a constructor for all required arguments ([check out the docs](https://projectlombok.org/features/Data)). The `@AllArgsConstructor` annotation generates an constructor for all fields in the class. I really like it because it de-clutters the data model classes and saves some time generating all of the 'ceremony' methods. 
+
+Next create a `HatService.java`. This is in lieu of using JPA and an actual database to persist the Hat object instances. It's just for demonstration purposes and simply creates two hats --  a large, red sombrero and a smaller blue beanie -- and serves them by ID.  
+
+`src/main/java/com/example/contract/HatService.java`
+
+```java
+package com.example.contract;
+
+import org.springframework.stereotype.Service;
+
+import java.util.HashMap;
+import java.util.Map;
+
+@Service
+public class HatService {
+    private final Map<Long, Hat> hatMap;
+    public HatService() {
+        hatMap = new HashMap<>();
+        hatMap.put(1L, new Hat(1L, "Sombrero", 30L, "red"));
+        hatMap.put(2L, new Hat(2L, "Beanie", 5L, "blue"));
+    }
+    Hat findHatById(Long id) {
+        return hatMap.get(id);
+    }
+}
+```
+
+Lastly, create a `HatController.java` that is what exposes your single resource endpoint. 
+
+`src/main/java/com/example/contract/HatController.java`
+
+```java
+package com.example.contract;
+
+import org.springframework.http.HttpStatus;
+import org.springframework.web.bind.annotation.GetMapping;
+import org.springframework.web.bind.annotation.PathVariable;
+import org.springframework.web.bind.annotation.RestController;
+import org.springframework.web.server.ResponseStatusException;
+
+@RestController
+public class HatController {
+    private final HatService hatService;
+
+    public HatController(HatService hatService) {
+        this.hatService = hatService;
+    }
+
+    @GetMapping("hat/{id}")
+    public Hat findHatById(@PathVariable("id") Long id) {
+        Hat hat = hatService.findHatById(id);
+        if (hat == null) {
+            throw new ResponseStatusException(
+                    HttpStatus.NOT_FOUND, "Not Found"
+            );
+        }
+        return hat;
+    }
+}
+```
+
+The HatController uses Spring's dependency injection to get the HatService as a bean. It then uses the service to lookup hats by ID and serve them to the consumer of the REST service endpoint at `/hat/{id}`. 
```

Comment:
Would it make sense to add a link to what is meant by 'bean'? Something like this: https://www.baeldung.com/spring-bean
Obviously I'm not your target reader, but for a moment I harked back to the blue beanie... Thanks!

Reply:
It's a pretty basic concept in Spring. It's sort of a foundational idea. I dunno if we're getting into that level of tutorial.

Reply:
Maybe say "as a Spring bean?" Or just remove "as a bean." It's prob not necessary anyway.

Reply:
I'll leave it as you wrote it.